### ISSUE TITLE
Argument for index file

### ISSUE THREAD

**Issue by AlistairNWard** [samtools/samtools#485]
Would it be possible to include a command line argument for the BAM index? Currently, assuming the index has FILE.bam.bai extension fails when tools, e.g. Picard have been used to generate an index of the form FILE.bai. If there is a command line argument, we can assume FILE.bam.bai in its absence, or use the value specified otherwise.

This is a significant overhead at the moment, so would really appreciate this simple fix.

**Comment by mcshane** [comment]
This will be possible again in the 1.3 release (8f60492d1a81b362cf7630ee3ab8669fd21420e0) via:
```
samtools index /path/to/bam /path/to/index
```

**Comment by AlistairNWard** [comment]
This doesn't resolve the issue. This lets me create any index but not use the index. I need to run a command:

samtools view /path/to/bam region

This is only possible for an indexed BAM and the assumption is that the index is FILE.bam.bai. If the index is FILE.bai, I cannot view this region. I need to be able to use the argument:

samtools view -x FILE.bai FILE.bam region

so that the index file can still be used.

**Comment by mcshane** [comment]
samtools will always look for look for both `FILE.bam.bai` and `FILE.bai`.
```
$ samtools view test/dat/test_input_1_a.bam ref1:7-7
[main_samview] random alignment retrieval only works for indexed BAM or CRAM files.
```
```
$ samtools index test/dat/test_input_1_a.bam test/dat/test_input_1_a.bai
$ samtools view test/dat/test_input_1_a.bam ref1:7-7
r001	163	ref1	7	30	8M4I4M1D3M	=	37	39	TTAGATAAAGAGGATACTG	*	XX:B:S,12561,2,20,112	YY:i:100	RG:Z:fish	PG:Z:colt
```

**Comment by AlistairNWard** [comment]
Thanks for the help Shane, I'm actually thinking of a different scenario. A use case we have is streaming a BAM file through samtools via https, where the bam file name is obfuscated. Samtools then assumes an index of the same name with .bai or .bam.bai appended. But given the obfuscation, this file may not exist. So, we would want to supply both the BAM and index files on the command line. Does this make sense?

**Comment by adamspargo** [comment]
Can you not obfuscate the name before you index the BAM, so that the index 
has the same name? Or maybe I'm missing something.

On Thu, 12 Nov 2015, Alistair Ward wrote:

> 
> Thanks for the help Shane, I'm actually thinking of a different scenario. A use case we have is streaming a BAM file
> through samtools via https, where the bam file name is obfuscated. Samtools then assumes an index of the same name
> with .bai or .bam.bai appended. But given the obfuscation, this file may not exist. So, we would want to supply both
> the BAM and index files on the command line. Does this make sense?
> 
> —
> Reply to this email directly or view it on GitHub.[ANTFDh8JQxuN9OxZ3CZU0D6U8OS2q_8jks5pFKxQgaJpZM4GghQU.gif]
> 
> 
>


---------------------------------------------------------
Dr Adam Spargo                    room: N312
Vertebrate Resequencing           email: aws@sanger.ac.uk
Wellcome Trust Sanger Institute   Tel: +44 (0)1223 834244
Hinxton, Cambridge CB10 1SA       ext: 8633


-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Comment by jmarshall** [comment]
That is indeed a different scenario from the one you have described previously…

Samtools and Picard and other similar tools have always looked alongside BAM files for similarly-named corresponding index files.  This has always been a limitation, but one that most people have been happy enough with, as it is a trade of flexibility for simplicity.

I have long considered implementing something like `foo.bam:bar.bai` to allow explicitly specifying the index file name (it needs to be bound to the filename argument rather than being a separate option for the sake of commands such as `mpileup` that work on multiple files), but it has never been a priority as there is little to no demand.

Other people who obfuscate filenames like this often manage to use the same obfuscated name for files as tightly conjoined as these two are.  Would it not be possible for you to apply the same obfuscating transformation to _foo.bam.bai_ as you do to _foo.bam_?

**Comment by AlistairNWard** [comment]
All of the obfuscated files that we create do as you say, so we currently
don't have the problem I'm outlining. I was bringing this up because my
experience is that everything that can happen, does happen! So I want to be
sure that as we increase our base of users, all of whom have their own
methods of obfuscation (and don't necessarily have an understanding of the
underlying files), we will be able to handle all cases. I thought that the
change would be straightforward, so we would be able to guarantee support
down the line.


Alistair Ward
Director, Research and Science
Department of Human Genetics
USTAR Center for Genetic Discovery
Eccles Institute of Human Genetics
University of Utah School of Medicine​
15 North 2030 East, Room 7140
Salt Lake City, UT 84112-5330

Co-founder, Frameshift labs, LLC

**Comment by agarwal** [comment]
I'd like to second @AlistairNWard's request. It should be possible to override default values, and the fix in this case seems simple enough. In complex workflow systems, you sometimes want to store files differently, and it's a bit annoying to not be able to specify the path.

**Comment by lh3** [comment]
The original question is about loading `FILE.bai` which according to @mcshane has been resolved.

As to the new request, the most relevant part is @AlistairNWard's remote access. Suppose we have already downloaded all 1000g *.bam.bai files, renamed them to `sha1(bam_URL).bai`, where function `sha1()` computes the SHA1 of a string, and are storing them in a local directory. Now when we want to query a remote 1000g BAM, it would be handy to retrieve data with `samtools view --index sha1(bam_URL).bai bam_URL`, but with the current samtools, we have to rename `sha1(bam_URL).bai` to a proper name for samtools view to work – not a deal breaker, but less convenient.

**Comment by AlistairNWard** [comment]
Is there any desire to resolve this issue within samtools?

**Comment by jmarshall** [comment]
It would appear that there is a moderate desire…

It is not something that the maintainers are likely to prioritise themselves anytime soon.  While @lh3 describes a reasonable use case and @agarwal seems to have actually encountered this (and we have other reports, of people who worked around their self-imposed filename obfuscation with symlink farms), @AlistairNWard last reported that this is for him merely a theoretical problem that he has not in fact encountered in the wild, and the fact is that looking for a BAI/CSI with a corresponding filename alongside a BAM file has sufficed for samtools, Picard, and other tools for a long time.

So I wonder whether you are also pressing for similar functionality in Picard, IGV, and all the other tools that use the same index file naming convention?

Having said that, while not prioritising working on this ourselves, we would entertain pull requests adding this functionality:

1. We are not interested in adding a simple `--index FNAME` option, because this cannot be extended to commands that take multiple input files.

2. Many commands take a file-of-filenames list file (e.g. `mpileup -b FOFN`), which could be extended to a tab-delimited file of `FNAME [\t INDEX_FNAME]` lines.  This would provide a way of specifying explicit index filenames attached to specific BAM/CRAM files even though that still couldn't be specified directly on the command line.

    To extend this to all samtools subcommands, either file-of-filenames options could be added to all subcommands, or perhaps file-of-filenames options on multi-input commands and `--index FNAME` options on simpler commands would suffice (though this approach is inferior).

3. Alternatively, a general `--index-separator CHAR` option could be added which would enable splitting of filename arguments into file / index-file at the user-specified `CHAR`.  It would probably want to operate wherever sequencing file filenames appear, so both as command line arguments and option arguments, and within files-of-filenames.

    So `--index-separator :` would switch on `foo.bam:bar.bai` as in my comment above, and `--index-separator "\t"` could switch on tab-separated file-of-filename processing as in (2) above.  Being user-specified would mean that no particular character became harder to use in filenames and being off by default would mean that the simple approach was still the default.

**Comment by AlistairNWard** [comment]
Thanks for responding to this John. I actually mentioned this issue again
since it isn't theoretical anymore. We are working with a number of
collaborators and we do need this functionality in samtools one way or
another. I am not pressing for this functionality in other tools (e.g.
Picard, IGV) since they are not tools that fall into our pipeline in the
same way and so don't need the same treatment. I can see that for others,
there would be an issue with other tools, but for me it is only samtools
that is pressing.

As for your implementation suggestions, I think what you listed in 3. seems
perfectly reasonable.

Thanks for discussing this again,

Al


Alistair Ward, PhD
Director, Research and Science
Department of Human Genetics
USTAR Center for Genetic Discovery
Eccles Institute of Human Genetics
University of Utah School of Medicine​
15 North 2030 East, Room 7140
Salt Lake City, UT 84112-5330

Co-founder, Frameshift labs, LLC

**Comment by chmille4** [comment]
Has there been any further movement on this? 

I'm running into this issue with trying to use data from DNANexus. They obfuscate each link independently so that you can't guess URLs. This makes a lot of sense, for example I've given a collaborator a link to Sample1.bam but I don't want to give them link to Sample2.bam, which would be trivial to guess if the obfuscation was consistent across all links.  Unfortunately, this means the .bai url can't be guessed just by adding .bai to the bam url.

**Comment by kdaily** [comment]
Would like to add a vote for implementing this. We develop a client that interacts with synapse.org. Files are stored in Amazon S3, but when the client retrieves them (or, information such as their path) it uses a similar technique with MD5 to prevent collisions, allow versioning, etc. Our users would like to be able to specify the index file separately since it will physically reside in a different directory.

**Comment by smrgit** [comment]
@jmarshall your last thoughtful response on this dates back over 2 years ... what are your current thoughts on this?  afaik tools like Picard / htsjdk do allow for the case where the BAI filename cannot be trivially inferred from the BAM filename, but that this remains a limitation in samtools

**Comment by jmarshall** [comment]
Me, my thoughts are unchanged. I continue to think any “just add a --index-file option” approach is a chimera and that the tab-separated fofn etc approach has legs. However samtools's current maintainers may be less committed to this “solve the problem for all scenarios” policy and to tasteful UIs, and less resistant to adding ‑‑index-file options that could incompletely paste over the cracks of small parts of the problem.

(I don't understand @chmille4's message re DNAnexus: it seems ridiculous for DNAnexus not to obfuscate the BAM and its associated BAI the same way. Sure, obfuscate sample1 and sample2 differently, but I don't see what's gained by doing the BAIs differently.)

I am unaware of any Picard options to let the command-line tools look for index files other than by trivially adding `.bai` etc to the filename, as samtools does. If you have information otherwise, please tell us about the specific Picard options.

**Comment by aloraine** [comment]
I'm writing a web service that retrieves BAM data from URLs by running samtools commands. This downloads the index file onto the web service host. However, many times, the index files have the same name. The --index-file option describe above would be very useful. Has this been implemented for samtools? Any suggestions on how to do that? I don't know what the code looks like (yet) but it seems like it would not be too hard to implement this type of thing.

**Comment by daviesrob** [comment]
This should be fixed now that #978 has been merged.

**Comment by kdaily** [comment]
Huzzah!
